mbuild/conversion.py

```diff
@@ -1109,13 +1117,94 @@ def save(
         structure.save(filename, overwrite=overwrite, **kwargs)
 
 
+def pull_residues(
+    compound,
+    segment_level=0,
+    include_base_level=False,
+):
+    """Pull residues from a Compound object.
+
+    Search class instance for completed compounds based on the number of
+    particles and bonds. If for example and peptide chain with three
+    individual peptides that were connected and hydrated with thirty water
+    molecules, the list will contain 31 residues. However, if the water and
+    the individual peptides should be labeled as individual residues, set
+    ``segment_level==1`` to receive a list with 33 residues. Depending on
+    the method used to assemble the peptides, this procedure may continue to
+    set ``segment_level=2`` and breakdown the peptide into functional groups.
+    Setting ``include_base_level==True`` will allow this procedure to
+    function with coarse-grained systems, while the default behavior will
+    end a search at the level before atoms are obtained.
+
+    Parameters
+    ----------
+    compound : obj
+        An instance of :class:`mbuild.compound.Compound`
+    segment_level : int, optional, default=0
+        Level of full residue architecture to be identified
+    include_base_level : bool, optional, default=False
+        Set whether a search should continue if the list of children are single particles.
+
+    Returns
+    -------
+    residues : list
+        List of residue names (str).
+    """
+    value_list = []
+    no_grandchildren = [not child.children for child in compound.children]
+    balanced_bonds = (
```

Comment:
I think ring structures will fail the check for compound.n_particles - compound.n_bonds == 1. 

Here are some examples comparing the behavior:
```python 
compound = mb.load("C1=CC=CC=C1", smiles=True) #benzene
print(compound, len(pull_residues(compound)))

compound = mb.Compound([mb.load("C1=CC=CC=C1", smiles=True), mb.load("C1=CC=CC=C1", smiles=True)])
print(compound, len(pull_residues(compound)))

compound = mb.Compound([mb.load("C1=CC=CC=C1", smiles=True), mb.load("C1=CC=CC=C1", smiles=True)])
compound.add_bond([compound.children[0][0], compound.children[1][0]])
print(compound, len(pull_residues(compound)))

compound = mb.load("CCC", smiles=True)
print(compound, len(pull_residues(compound)))

compound = mb.Compound([mb.load("CCC", smiles=True), mb.load("CCC", smiles=True)])
print(compound, len(pull_residues(compound)))

compound = mb.Compound([mb.load("CCC", smiles=True), mb.load("CCC", smiles=True)])
compound.add_bond([compound.children[0][0], compound.children[1][0]])
print(compound, len(pull_residues(compound)))
```
Output
```
<Compound 12 particles, 12 bonds, non-periodic, id: 140616982879824> 1
<Compound 24 particles, 24 bonds, non-periodic, id: 140616984251504> 2
<Compound 24 particles, 25 bonds, non-periodic, id: 140616981578368> 2
<Compound 11 particles, 10 bonds, non-periodic, id: 140616983072832> 1
<Compound 22 particles, 20 bonds, non-periodic, id: 140616981578368> 2
<Compound 22 particles, 21 bonds, non-periodic, id: 140616982976784> 1
```

As you can see, the issue arises in the difference between the 3rd and 6th print statements here, which are bonded subcompounds. If there is a ring structure, it treats those as two different molecules instead of one, since balanced bonds is set to False, whereas in the 6th example, with two bonded "CCC" moieties, it treats that as a single molecule, as it should.

The solution could be to use compound.bond_graph.connected_components here to check the structure.